`test_worker_thread_context_not_leaked` sometimes fails

cc @EmmanuelNiyonshuti 

I haven't dug into this yet but the test is flakey. I'm opening this instead of simply commenting in the flakey-tests thread so we can discuss!

See https://github.com/A5rocks/trio/actions/runs/30208152029 and https://github.com/A5rocks/trio/actions/runs/29643389046

> > Wait, it actually fails on CPython!!! https://github.com/A5rocks/trio/actions/runs/30276990757/job/90013769036#step:4:1300
> 
> thats weird! I was testing it earlier today.

I'm not able to see the test in question failing; how are you able to see that? All I get from the linked run is:  `Error: Process completed with exit code 1`

I see this in the logs:
```
  ================================== FAILURES ===================================
  ____________________ test_worker_thread_context_not_leaked ____________________
  
      async def test_worker_thread_context_not_leaked() -> None:
          # Regression test for: https://github.com/python-trio/trio/issues/3472
      
          class Foo:
              pass
      
          def sync_fn() -> None:
              pass
      
          cvar: contextvars.ContextVar[Foo] = contextvars.ContextVar("cvar")
          contextval = Foo()
          ref = weakref.ref(contextval)
          cvar.set(contextval)
          await to_thread_run_sync(sync_fn)
          cvar.set(Foo())
      
          del contextval
          if sys.implementation.name == "pypy":
              import gc
              gc.collect()
      
  >       assert ref() is None
  E       AssertionError: assert <trio._tests.test_threads.test_worker_thread_context_not_leaked.<locals>.Foo object at 0x000001932E4CB740> is None
  E        +  where <trio._tests.test_threads.test_worker_thread_context_not_leaked.<locals>.Foo object at 0x000001932E4CB740> = <weakref at 0x000001932E4FADE0; to 'Foo' at 0x000001932E4CB740>()
  
  C:\hostedtoolcache\windows\Python\3.12.10\x64\Lib\site-packages\trio\_tests\test_threads.py:711: AssertionError
  ------------- generated xml file: D:\a\trio\trio\test-results.xml -------------
  ============================ slowest 10 durations =============================
  0.01s call     _tests/test_threads.py::test_worker_thread_context_not_leaked
  
```

(notably I changed `gc.collect` to only happen on pypy, so this is possible if there's a refcycle or something. but.... there shouldn't be a refcycle that only happens like 1% of the time?)

--- a/src/trio/_core/_thread_cache.py
+++ b/src/trio/_core/_thread_cache.py
@@ -188,6 +188,13 @@ class WorkerThread(Generic[RetT]):
             if set_os_thread_name:
                 set_os_thread_name(self._thread.ident, self._default_name)
 
+        # Without this, this thread may get rescheduled after `deliver`
+        # and only return after the kernel gets around to it. That's an
+        # arbitrary amount of time during which we keep anything
+        # referred to in the function alive, like contextvars. That's
+        # surprising!
+        del fn
+
         # Tell the cache that we're available to be assigned a new
         # job. We do this *before* calling 'deliver', so that if
         # 'deliver' triggers a new job, it can be assigned to us

--- a/src/trio/_tests/test_threads.py
+++ b/src/trio/_tests/test_threads.py
@@ -1,7 +1,6 @@
 from __future__ import annotations
 
 import contextvars
-import gc
 import queue as stdlib_queue
 import re
 import sys
@@ -29,7 +28,7 @@ from .. import (
     sleep_forever,
 )
 from .._core._tests.test_ki import ki_self
-from .._core._tests.tutil import slow
+from .._core._tests.tutil import gc_collect_harder, slow
 from .._threads import (
     active_thread_count,
     current_default_thread_limiter,
@@ -705,7 +704,8 @@ async def test_worker_thread_context_not_leaked() -> None:
     cvar.set(Foo())
 
     del contextval
-    gc.collect()
+    if sys.implementation.name == "pypy":
+        gc_collect_harder()
 
     assert ref() is None
 

